### PR TITLE
gh-120521: clarify except* documentation to allow tuples

### LINKED ISSUE #120521: Clarify documentation for except*
# Documentation

The [documentation for except*](https://docs.python.org/3/reference/compound_stmts.html#except-clause) states

> An except* clause must have a matching type, and this type cannot be a subclass of [BaseExceptionGroup](https://docs.python.org/3/library/exceptions.html#BaseExceptionGroup).

This incorrectly implies that there can only be one matching type, when tuples of types are allowed in except*.

I propose to change this to:

> An except* clause must have a matching type or a tuple of matching types, which cannot be a subclass of [BaseExceptionGroup](https://docs.python.org/3/library/exceptions.html#BaseExceptionGroup).

This has been discussed in the forum: https://discuss.python.org/t/clarifying-typing-behavior-for-exception-handlers-except-except/55060

### Linked PRs
* gh-120523
* gh-120750
* gh-120751

### PR DESCRIPTION
**Pull request by yangdanny97**
fixes #120521

Discussion link: https://discuss.python.org/t/clarifying-typing-behavior-for-exception-handlers-except-except/55060/2

* Issue: gh-120521

----
📚 Documentation preview 📚: https://cpython-previews--120523.org.readthedocs.build/

### CODE REVIEW

File: Doc/reference/compound_stmts.rst [hunk 1]
```diff
@@ -378,8 +378,8 @@ exception group with an empty message string. ::
    ...
    ExceptionGroup('', (BlockingIOError()))
 
-An :keyword:`!except*` clause must have a matching type,
-and this type cannot be a subclass of :exc:`BaseExceptionGroup`.
+An :keyword:`!except*` clause must have a matching type or a tuple of matching
```
Comment (JelleZijlstra) [inline Doc/reference/compound_stmts.rst]: It might be worth expanding this further:

- The docs for `except` (the previous section) never say that each type must be a subtype of `BaseException`. That's probably worth adding.
- One important difference between `except` and `except*` is that the former allows "bare" except clauses (just `except:`), but `except*` doesn't. I think that's worth mentioning in a sentence by itself.
- We don't say what happens if the type is a subclass of BaseExceptionGroup.

So I'd suggest something like this:

```suggestion
Unlike :keyword:`except` clauses, :keyword:`!except*` clauses must have a matching
expression; `except*:` is not valid syntax. When the handler is evaluated, the
expression must evaluate to a type or a tuple containing types. Each type must be a
subclass of :exc:`BaseException` and must not be a subclass of :exc:`BaseExceptionGroup`.
```

Comment (yangdanny97) [reply Doc/reference/compound_stmts.rst]: Okay, I've added a mention that the runtime will raise a TypeError if the type is wrong, and updated the `except` section to have similar language.

Should the latter be split into a separate PR, or should I just update the title/summary of this PR?

File: Doc/reference/compound_stmts.rst [hunk 2]
```diff
@@ -245,13 +245,17 @@ handler is started. This search inspects the :keyword:`!except` clauses in turn
...
-that expression is evaluated, and the clause matches the exception
-if the resulting object is "compatible" with the exception.  An object is
-compatible with an exception if the object is the class or a
-:term:`non-virtual base class <abstract base class>` of the exception object,
-or a tuple containing an item that is the class or a non-virtual base class
-of the exception object.
+
+For an :keyword:`!except` clause with an expression, the
+expression must evaluate to a type or a tuple containing types. Each type must
+be a subclass of :exc:`BaseException`, otherwise the runtime will raise a
```
Comment (JelleZijlstra) [inline Doc/reference/compound_stmts.rst]:
```suggestion
be a subclass of :exc:`BaseException`; otherwise, the runtime will raise a
```
Run-on sentence

File: Doc/reference/compound_stmts.rst [hunk 3]
```diff
@@ -378,8 +382,13 @@ exception group with an empty message string. ::
    ...
    ExceptionGroup('', (BlockingIOError()))
 
-An :keyword:`!except*` clause must have a matching type,
-and this type cannot be a subclass of :exc:`BaseExceptionGroup`.
+Unlike :keyword:`except` clauses, :keyword:`!except*` clauses must have a matching
+expression; ``except*:`` is not valid syntax. When the handler is evaluated, the
+expression must evaluate to a type or a tuple containing types. Each type must
+be a subclass of :exc:`BaseException` and must not be a subclass of
+:exc:`BaseExceptionGroup`, otherwise the runtime will raise a :exc:`TypeError`
```
Comment (JelleZijlstra) [inline Doc/reference/compound_stmts.rst]:
```suggestion
:exc:`BaseExceptionGroup`; otherwise, the runtime will raise a :exc:`TypeError`
```
Same

**Review by JelleZijlstra — COMMENTED** [review]
This looks right, cc @iritkatriel for `except*`.

File: Doc/reference/compound_stmts.rst [hunk 4]
```diff
@@ -245,13 +245,17 @@ handler is started. This search inspects the :keyword:`!except` clauses in turn
...
-that expression is evaluated, and the clause matches the exception
-if the resulting object is "compatible" with the exception.  An object is
-compatible with an exception if the object is the class or a
-:term:`non-virtual base class <abstract base class>` of the exception object,
-or a tuple containing an item that is the class or a non-virtual base class
-of the exception object.
+
+For an :keyword:`!except` clause with an expression, the
+expression must evaluate to a type or a tuple containing types. Each type must
+be a subclass of :exc:`BaseException`; otherwise, the runtime will raise a
```
Comment (iritkatriel) [inline Doc/reference/compound_stmts.rst]: It can be BaseException itself. Can this be worded to make that clear?

File: Doc/reference/compound_stmts.rst [hunk 5]
```diff
@@ -245,13 +245,17 @@ handler is started. This search inspects the :keyword:`!except` clauses in turn
...
+:exc:`TypeError` when the handler is evaluated.
+
+The clause matches the exception if the resulting object is "compatible" with
```
Comment (iritkatriel) [inline Doc/reference/compound_stmts.rst]:
```suggestion
The clause matches an exception if the resulting object is "compatible" with
```

Comment (iritkatriel) [inline Doc/reference/compound_stmts.rst]: Since you moved this sentence to a new paragraph, I don't think it's clear what "the resulting object" refers to.

File: Doc/reference/compound_stmts.rst [hunk 6]
```diff
@@ -378,8 +382,13 @@ exception group with an empty message string. ::
...
    ExceptionGroup('', (BlockingIOError()))
 
-An :keyword:`!except*` clause must have a matching type,
-and this type cannot be a subclass of :exc:`BaseExceptionGroup`.
+Unlike :keyword:`except` clauses, :keyword:`!except*` clauses must have a matching
+expression; ``except*:`` is not valid syntax. When the handler is evaluated, the
+expression must evaluate to a type or a tuple containing types. Each type must
+be a subclass of :exc:`BaseException` and must not be a subclass of
+:exc:`BaseExceptionGroup`; otherwise, the runtime will raise a :exc:`TypeError`
+when the handler is evaluated.
```
Comment (iritkatriel) [inline Doc/reference/compound_stmts.rst]: Can we not repeat this twice? It is confusing because people assume it will somehow be different from the except case.

File: Doc/reference/compound_stmts.rst (hunk 5 above)
Comment (yangdanny97) [reply Doc/reference/compound_stmts.rst]: added a clarification to "resulting object from evaluating the clause"

File: Doc/reference/compound_stmts.rst (hunk 4 above)
Comment (yangdanny97) [reply Doc/reference/compound_stmts.rst]: clarified to "must be :exc:`BaseException` or a subclass of :exc:`BaseException`"

same for the BaseExceptionGroup mention in the except* section

File: Doc/reference/compound_stmts.rst (hunk 6 above)
Comment (yangdanny97) [reply Doc/reference/compound_stmts.rst]: Replaced the duplicate part with "In addition to the requirements for matching expressions in except clauses"

```
In addition to the
requirements for matching expressions in :keyword:`except` clauses, the
types in the matching expression of an :keyword:`!except*` clause cannot
be :exc:`BaseExceptionGroup` or a subclass of :exc:`BaseExceptionGroup`;
otherwise, the runtime will raise a :exc:`TypeError` when the handler is
evaluated.
```

**Review by offensive-vk — APPROVED** [review]
Nice explanation!

**Comment by iritkatriel** [comment]
I think this PR should just resolve the issue it is associated with. As it is now it is doing a lot more than that. 
For example, the fact that an incorrect exception expression results in TypeError at runtime is an implementation detail, it is not currently (as far as I know) documented in the language spec. 

Can you limit the PR to just resolve the problem raised in the issue?

**Comment by yangdanny97** [comment]
> I think this PR should just resolve the issue it is associated with. As it is now it is doing a lot more than that. For example, the fact that an incorrect exception expression results in TypeError at runtime is an implementation detail, it is not currently (as far as I know) documented in the language spec.
> 
> Can you limit the PR to just resolve the problem raised in the issue?

I'll remove the `TypeError` part, though I think the rest of it is kind of difficult to untangle because the `except*` docs now references the `except` docs (with the "in addition to"/"unlike" wording), so it might make sense to keep the changes together.

File: Doc/reference/compound_stmts.rst [hunk 7]
```diff
@@ -245,13 +245,17 @@ handler is started. This search inspects the :keyword:`!except` clauses in turn
...
-that expression is evaluated, and the clause matches the exception
-if the resulting object is "compatible" with the exception.  An object is
-compatible with an exception if the object is the class or a
-:term:`non-virtual base class <abstract base class>` of the exception object,
-or a tuple containing an item that is the class or a non-virtual base class
-of the exception object.
+
+For an :keyword:`!except` clause with an expression, the
+expression must evaluate to a type or a tuple containing types. Each type must
+be :exc:`BaseException` or a subclass of :exc:`BaseException`.
```
Comment (iritkatriel) [inline Doc/reference/compound_stmts.rst]:
```suggestion
expression must evaluate to an exception type or a tuple of exception types.
```

File: Doc/reference/compound_stmts.rst [hunk 8]
```diff
@@ -245,13 +245,17 @@ handler is started. This search inspects the :keyword:`!except` clauses in turn
...
+
+The clause matches an exception if the resulting object from evaluating the
+clause is "compatible" with the exception. An object is compatible with an
+exception if the object is the class or a
+:term:`non-virtual base class <abstract base class>` of the exception
+object, or a tuple containing an item that is the class or a non-virtual base
+class of the exception object.
```
Comment (iritkatriel) [inline Doc/reference/compound_stmts.rst]:
```suggestion
The raised exception matches an :keyword:`!except` clause whose expression evaluates
to the class or a :term:`non-virtual base class <abstract base class>` of the exception object,
or it evaluates to a tuple that contains such a class.
```

File: Doc/reference/compound_stmts.rst [hunk 9]
```diff
@@ -378,8 +382,12 @@ exception group with an empty message string. ::
    ...
    ExceptionGroup('', (BlockingIOError()))
 
-An :keyword:`!except*` clause must have a matching type,
-and this type cannot be a subclass of :exc:`BaseExceptionGroup`.
+Unlike :keyword:`except` clauses, :keyword:`!except*` clauses must have a
+matching expression; ``except*:`` is not valid syntax. In addition to the
+requirements for matching expressions in :keyword:`except` clauses, the
+types in the matching expression of an :keyword:`!except*` clause cannot
+be :exc:`BaseExceptionGroup` or a subclass of :exc:`BaseExceptionGroup`.
```
Comment (iritkatriel) [inline Doc/reference/compound_stmts.rst]:
```suggestion
An :keyword:`!except*` clause must have a matching expression; it cannot be ``except*:``.
Furthermore, this expression cannot contain exception group types, because that would
have ambiguous semantics.
```

File: Doc/reference/compound_stmts.rst [hunk 10]
```diff
@@ -378,8 +382,12 @@ exception group with an empty message string. ::
    ...
    ExceptionGroup('', (BlockingIOError()))
 
-An :keyword:`!except*` clause must have a matching type,
-and this type cannot be a subclass of :exc:`BaseExceptionGroup`.
+Unlike :keyword:`except` clauses, :keyword:`!except*` clauses must have a
+matching expression; ``except*:`` is not valid syntax. In addition to the
```
Comment (iritkatriel) [inline Doc/reference/compound_stmts.rst]: (I wouldn't say it's invalid syntax when it generates a runtime error).

File: Doc/reference/compound_stmts.rst (hunk 9 above)
Comment (iritkatriel) [inline Doc/reference/compound_stmts.rst]: (Above we talk about "exception group" when we refer to BaseException and its subclasses).

**Review by iritkatriel — COMMENTED** [review]
I think it became a bit wordy. I tried to reword it more concisely.

File: Doc/reference/compound_stmts.rst (hunk 10 above)
Comment (JelleZijlstra) [reply Doc/reference/compound_stmts.rst]: It's a SyntaxError:
```
>>> try: x
... except *: pass
  File "<stdin>", line 2
    except *: pass
            ^
SyntaxError: expected one or more exception types
```

Comment (iritkatriel) [reply Doc/reference/compound_stmts.rst]: Ok, that one is but not this one:

```
>>> try: 1/0
... except* BaseExceptionGroup: pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ZeroDivisionError: division by zero

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
TypeError: catching ExceptionGroup with except* is not allowed. Use except instead.
```

File: Doc/reference/compound_stmts.rst (hunk 9 above)
Comment (yangdanny97) [reply Doc/reference/compound_stmts.rst]: Sorry, I'm confused about this comment - is it addressed in your suggestions already?

The other suggestions on this PR which I've merged in seem to have removed all references to BaseException and BaseExceptionGroup classnames in this doc. 

Feels like explicitly mentioning the class names would be more clear, wouldn't it?

Comment (iritkatriel) [reply Doc/reference/compound_stmts.rst]: Yes, this comment is addressed.

Since we talk about "exception group" in the previous paragraphs, assuming the reader knows what we mean there, I don't think specifying the type here would add clarity.

**Review by iritkatriel — APPROVED** [review]
LGTM, I'll let @JelleZijlstra have another look.

File: Doc/reference/compound_stmts.rst [hunk 11]
```diff
@@ -245,13 +245,12 @@ handler is started. This search inspects the :keyword:`!except` clauses in turn
...
-compatible with an exception if the object is the class or a
-:term:`non-virtual base class <abstract base class>` of the exception object,
-or a tuple containing an item that is the class or a non-virtual base class
-of the exception object.
+
+For an :keyword:`!except` clause with an expression, the
+expression must evaluate to an exception type or a tuple of exception types.
+The raised exception matches an :keyword:`!except` clause whose expression evaluates
+to the class or a :term:`non-virtual base class <abstract base class>` of the exception object,
+or it evaluates to a tuple that contains such a class.
```
Comment (JelleZijlstra) [inline Doc/reference/compound_stmts.rst]:
```suggestion
or to a tuple that contains such a class.
```
I think this makes for a better sentence structure.

File: Doc/reference/compound_stmts.rst [hunk 12]
```diff
@@ -245,13 +245,12 @@ handler is started. This search inspects the :keyword:`!except` clauses in turn
...
-if the resulting object is "compatible" with the exception.  An object is
-compatible with an exception if the object is the class or a
-:term:`non-virtual base class <abstract base class>` of the exception object,
-or a tuple containing an item that is the class or a non-virtual base class
-of the exception object.
+
+For an :keyword:`!except` clause with an expression, the
+expression must evaluate to an exception type or a tuple of exception types.
+The raised exception matches an :keyword:`!except` clause whose expression evaluates
+to the class or a :term:`non-virtual base class <abstract base class>` of the exception object,
```
Comment (JelleZijlstra) [inline Doc/reference/compound_stmts.rst:252]: This is a pre-existing problem and doesn't have to be solved in this PR, but I don't think linking to the definition of "abstract base class" here is great.

**Review by JelleZijlstra — APPROVED** [review]
Looks good, just a wording suggestion

**Review by JelleZijlstra — APPROVED** [review]
